### ISSUE TITLE
Add K Medoids Estimator to tf canned estimators

### ISSUE THREAD

**Issue by tambulkar** [tensorflow/tensorflow#34964]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
https://en.wikipedia.org/wiki/K-medoids
Basically same as https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/kmeans.py except when updating centroid use median rather than mean.

**Will this change the current api? How?**
No, it should just add functionality

**Who will benefit with this feature?**
Anyone who wants to use K-medoids clustering

**Any Other info.**

**Comment by tanzhenyu** [comment]
We're not actively developing estimator models, for future development, this should go into Keras premade models. And contributions welcome

**Comment by Byambaa0325** [comment]
Hello, can I work on it? I would like to try it out.

**Comment by tanzhenyu** [comment]
> Hello, can I work on it? I would like to try it out.

Of course. Can you start by sending out design ideas according to RFC rule?

**Comment by Byambaa0325** [comment]
> > Hello, can I work on it? I would like to try it out.
> 
> Of course. Can you start by sending out design ideas according to RFC rule?

Awesome, I'll prepare for a request. It is my first time, and I am a little lost:

- Can I get early feedback on the RFC here?
- Should the K Mediods better be added to the [tf canned estimator](https://www.tensorflow.org/tutorials/estimator/premade) or the [keras premade models](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/keras/premade) as mentioned above? 

For the first case, I am planning to take heavy inspiration from the existing implementation of [KMeans](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/kmeans.py) as the models are fundamentally similar in most aspects and even sklearn k_mediod model uses some parts from its k_means model.

If the Keras premade model implementation is preferred, I am thinking of incremental development as I am quite unfamiliar with both for now. I would love some pointers on this one, if possible.

**Comment by tanzhenyu** [comment]
> > > Hello, can I work on it? I would like to try it out.
> > 
> > 
> > Of course. Can you start by sending out design ideas according to RFC rule?
> 
> Awesome, I'll prepare for a request. It is my first time, and I am a little lost:
> 
> * Can I get early feedback on the RFC here?
> * Should the K Mediods better be added to the [tf canned estimator](https://www.tensorflow.org/tutorials/estimator/premade) or the [keras premade models](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/keras/premade) as mentioned above?
> 
> For the first case, I am planning to take heavy inspiration from the existing implementation of [KMeans](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/kmeans.py) as the models are fundamentally similar in most aspects and even sklearn k_mediod model uses some parts from its k_means model.
> 
> If the Keras premade model implementation is preferred, I am thinking of incremental development as I am quite unfamiliar with both for now. I would love some pointers on this one, if possible.

In the RFC process, it is suggested to first email developers@tensorflow.org before writing up the entire doc :-)
Regarding implementation, yeah some common logic would be preferred, but we don't have any immediate implementation to export TF2 KMeans yet.

**Comment by Byambaa0325** [comment]
> > > > Hello, can I work on it? I would like to try it out.
> > > 
> > > 
> > > Of course. Can you start by sending out design ideas according to RFC rule?
> > 
> > 
> > Awesome, I'll prepare for a request. It is my first time, and I am a little lost:
> > 
> > * Can I get early feedback on the RFC here?
> > * Should the K Mediods better be added to the [tf canned estimator](https://www.tensorflow.org/tutorials/estimator/premade) or the [keras premade models](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/keras/premade) as mentioned above?
> > 
> > For the first case, I am planning to take heavy inspiration from the existing implementation of [KMeans](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/kmeans.py) as the models are fundamentally similar in most aspects and even sklearn k_mediod model uses some parts from its k_means model.
> > If the Keras premade model implementation is preferred, I am thinking of incremental development as I am quite unfamiliar with both for now. I would love some pointers on this one, if possible.
> 
> In the RFC process, it is suggested to first email [developers@tensorflow.org](mailto:developers@tensorflow.org) before writing up the entire doc :-)
> Regarding implementation, yeah some common logic would be preferred, but we don't have any immediate implementation to export TF2 KMeans yet.
Thanks for being friendly and helpful.

**Comment by rayporten** [comment]
we could change option or have discussion to find the answer.not my answer should be best.

**Comment by Suraj-Upadhyay** [comment]
Hii @tanzhenyu, I want to follow up on this issue and add the K-Medoids api. After reading all the previous comments and discussions here and on the [mailing list](https://groups.google.com/a/tensorflow.org/g/developers/c/Jj0SORWweho/m/6awItltCAwAJ), I think there are a few things that needs to be decided before we can proceed with this issue.

1. There is going to be a lot of similarities with the k-means api and before we could even think of code sharing we need to decide upon the status of [clustering_ops.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/clustering_ops.py) w.r.t to the TODO written [here](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/kmeans.py#L17).

2. If we don't want to resolve the above issue first then we should proceed with adding a class `KMedoids` in the [clustering_ops.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/clustering_ops.py) and an estimator class `KMedoidsEstimator` in the [canned directory](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned) just like its counterpart.

Though, I like the second method way more because of its simplicity and straightforwardness. Or, maybe I am completely wrong about this.

However, I suppose the core developers might have got a different and better design for this.

So, Could you please guide me through this ?

Also, I would like to know if I have to send an RFC for this or not.

**Comment by sachinprasadhs** [comment]
Hi, Estimator will be deprecated from Tensorflow 2.12 since it uses Tensorflow V1 behavior.
Could you please migrate from Estimator to Keras API which will enable you to perform the aforementioned tasks, such as [model building](https://www.tensorflow.org/guide/keras/custom_layers_and_models), gradient application, [training](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit), evaluation, and prediction.
Refer Migration guide here https://www.tensorflow.org/guide/migrate/migrating_estimator for more details. Thanks!
